### ISSUE TITLE
Folds are calculated before the tree-sitter tree is updated 

### ISSUE THREAD

**Issue by stsewd** [neovim/neovim#14977]
- `nvim --version`: nightly
- Operating system/version: N/A
- Terminal name/version: N/A
- `$TERM`: N/A

### Steps to reproduce using `nvim -u NORC`

This can be replicated when using `nvim_buf_attach` and using tree-sitter for indentation. An easy way to replicate this is

```vim
call plug#begin('~/.local/share/nvim/plugged')
Plug 'nvim-treesitter/nvim-treesitter', {'do': ':TSUpdate'}
call plug#end()

set foldmethod=expr
set foldexpr=nvim_treesitter#foldexpr()
lua <<EOF
require'nvim-treesitter.configs'.setup {
  ensure_installed = "all",
  highlight = {
    enable = true,
  },
}
EOF
```

And a file like this one

```json
{
  "glossary": {
    "title": "example glossary",
    "GlossDiv": {
      "title": "S",
      "GlossList": {
        "GlossEntry": {
          "ID": "SGML",
          "SortAs": "SGML",
          "GlossTerm": "Standard Generalized Markup Language",
          "Acronym": "SGML",
          "Abbrev": "ISO 8879:1986",
          "GlossDef": {
            "para": "A meta-markup language, used to create markup languages such as DocBook.",
            "GlossSeeAlso": [
              "GML",
              "XML"
            ]
          },
          "GlossSee": "markup"
        }
      }
    }
  }
}
```

- Open the file
- Delete one node (say the `GlossSeeAlso` key)
- Folding is broken now

### Actual behaviour

- TS has the current content, with let's say 5 folds
- One fold is deleted
- Neovim calls the fold function, it has 5 folds!
- TS is updated to have 4 folds
- Undo is triggered
- Neovim calls the fold function, it has 4 folds!
- TS is updated to have 5 folds

### Expected behaviour

- TS has the current content, with let's say 5 folds
- One fold is deleted
- TS is updated to have 4 folds
- Neovim calls the fold function, it has 4 folds!
- Undo is triggered
- TS is updated to have 5 folds
- Neovim calls the fold function, it has 4 folds!

Basically, the tree is updated first before invoking the fold expression.

I think this is something along these lines https://github.com/neovim/neovim/blob/75f758b993b842c4afd03fdaa0063b573fdf8751/src/nvim/change.c#L342-L344 but not sure

ref https://github.com/nvim-treesitter/nvim-treesitter/issues/1424

**Comment by whereistejas** [comment]
Hi, any updates on this?

**Comment by nyngwang** [comment]
Hi, any updates on this?
